### ISSUE TITLE
Performance Degradation for Variable-Length Batching on FA3 (H100)

### ISSUE THREAD

**Issue by leigao97** [Dao-AILab/flash-attention#1767]
I am benchmarking the FA3 kernel on an H100, and I observed a significant performance gap between uniform-length and variable-length batches. Specifically, scenarios where most requests are decode (q_len=1) and just one is a prefill (q_len=L) perform much worse than when all requests are decode-only.

- GPU: NVIDIA H100 SXM
- FA3 version: vLLM 0.9.1 complied FA3 binary (_vllm_fa3_C.abi3.so)
- FA3 usage: Pack q, k, and v by squeezing the batch size dim and q_len dim. Prepare the cumulative q_len and cumulative k_len meta data. Set rest arguments as default or None.
```
torch.ops._vllm_fa3_C.fwd(q, k, v, None, None, None, None,
                cu_q, cu_k, None, None, None,
                max_q, max_k,
                None, None, None, None, None, None,
                None, None, None,
                SCALE, True, -1, -1, 0.0,
                None, None, 0, None, 0)
```
- Model shape:
  - head_dim: 128
  - num_heads: 32
  - num_kv_heads: 8
- Batch configs:
  - decode_only: all B requests with q_len = 1, k_len = L
  - decode_dominate: (B - 1) requests with q_len = 1, 1 request with q_len = L. All B requests with k_len = L

Plots below show the measured latency across different B and L for two batch configs.

<img width="1800" height="750" alt="Image" src="https://github.com/user-attachments/assets/890a35e6-ff9b-475e-b1d3-92ecd09238b7" />

The results show that decode_dominate batch config performs much worse than decode_only batch config, though including only one prefill request in the batch. 

Could you please clarify why this happens? Is this an expected limitation of the current FA3 implementation, or is there room for optimization?

**Comment by leigao97** [comment]
To verify this performance degradation is unexpected, I also benchmarked the latency for prefilling a single request. The results are shown below:

<img width="630" height="525" alt="Image" src="https://github.com/user-attachments/assets/3998dbd8-3071-4787-8bf3-25b527e9dfe9" />

If we simply launch FA3 kernel two times for decode-only batch and single-prefill batch, the added latency should be smaller than that of the decode-dominate batch.

**Comment by tridao** [comment]
Can you post a short script to produce the benchmark numbers?

**Comment by leigao97** [comment]
Thank you for your reply. I am using the following script. I copied the` _vllm_fa3_C.abi3.so` lib from vllm. You can search it by `find ~/. -name "_vllm_fa3_C.abi3.so"` If you don't have vllm installed, you can also use `flash_attn_varlen_func`, after uncommenting it in the script. 

I also found that var len feature works well for MHA but not GQA.
```
import math
import itertools

import torch
import pandas as pd
import matplotlib.pyplot as plt
import _vllm_fa3_C
from flash_attn import flash_attn_varlen_func

# --------------------------------------------------------------------------------------
HEAD_DIM     = 128
NUM_HEADS    = 32
NUM_KV_HEADS = 8
DTYPE        = torch.float16
DEVICE       = "cuda"
SCALE        = 1.0 / math.sqrt(HEAD_DIM)
WARMUP_ITERS = 20
MEASURE_ITERS = 100

SEQ_LENS     = [512, 1024, 2048, 4096]
BATCH_SIZES  = [4, 8, 16, 32, 64]
SCENARIOS    = ["decode_only", "decode_dominate"]

# --------------------------------------------------------------------------------------
# Utility functions
def build_flat_qkv(q_lens, k_lens):
    """Pack variable-length requests into FA-3 flattened format."""
    q_list, k_list, v_list = [], [], []
    cu_q, cu_k = [0], [0]

    for q_len, k_len in zip(q_lens, k_lens):
        q = torch.randn(q_len, NUM_HEADS, HEAD_DIM, dtype=DTYPE, device=DEVICE)
        k = torch.randn(k_len, NUM_KV_HEADS, HEAD_DIM, dtype=DTYPE, device=DEVICE)
        v = torch.randn_like(k)
        q_list.append(q)
        k_list.append(k)
        v_list.append(v)
        cu_q.append(cu_q[-1] + q_len)
        cu_k.append(cu_k[-1] + k_len)

    return (
        torch.cat(q_list), torch.cat(k_list), torch.cat(v_list),
        torch.tensor(cu_q, device=DEVICE, dtype=torch.int32),
        torch.tensor(cu_k, device=DEVICE, dtype=torch.int32),
    )

def scenario_lengths(scenario, B, L):
    if scenario == "decode_only":
        return [1]*B, [L]*B
    if scenario == "decode_dominate":
        return [1]*(B-1) + [L], [L]*B
    raise ValueError(scenario)

def fa3_latency_ms(q, k, v, cu_q, cu_k, max_q, max_k):
    """Average kernel time (ms) over MEASURE_ITERS iterations."""
    evt_start, evt_end = torch.cuda.Event(True), torch.cuda.Event(True)

    # Warm-up
    for _ in range(WARMUP_ITERS):
        # flash_attn_varlen_func(q, k, v,
        #                         max_seqlen_q=max_q, cu_seqlens_q=cu_q,
        #                         max_seqlen_k=max_k, cu_seqlens_k=cu_k,
        #                         softmax_scale=SCALE, causal=True
        # )
        torch.ops._vllm_fa3_C.fwd(q, k, v, None, None, None, None,
                        cu_q, cu_k, None, None, None,
                        max_q, max_k,
                        None, None, None, None, None, None,
                        None, None, None,
                        SCALE, True, -1, -1, 0.0,
                        None, None, 0, None, 0)

    torch.cuda.synchronize()
    evt_start.record()
    for _ in range(MEASURE_ITERS):
        # flash_attn_varlen_func(q, k, v,
        #                         max_seqlen_q=max_q, cu_seqlens_q=cu_q,
        #                         max_seqlen_k=max_k, cu_seqlens_k=cu_k,
        #                         softmax_scale=SCALE, causal=True
        # )
        torch.ops._vllm_fa3_C.fwd(q, k, v, None, None, None, None,
                        cu_q, cu_k, None, None, None,
                        max_q, max_k,
                        None, None, None, None, None, None,
                        None, None, None,
                        SCALE, True, -1, -1, 0.0,
                        None, None, 0, None, 0)
    evt_end.record()
    torch.cuda.synchronize()
    return evt_start.elapsed_time(evt_end) / MEASURE_ITERS

# --------------------------------------------------------------------------------------
# Benchmark loop
records = []
for L, B, scenario in itertools.product(SEQ_LENS, BATCH_SIZES, SCENARIOS):
    q_lens, k_lens = scenario_lengths(scenario, B, L)
    q, k, v, cu_q, cu_k = build_flat_qkv(q_lens, k_lens)
    measured = fa3_latency_ms(q, k, v, cu_q, cu_k, max(q_lens), max(k_lens))

    records.append(dict(seq_len=L, batch=B, scenario=scenario, latency_ms=measured))
    print(f"{scenario:<17} | B={B:<2} L={L:<4}  meas={measured:7.3f} ms")

df = pd.DataFrame(records)

# --------------------------------------------------------------------------------------
# Scatter plots for decode_only and decode_dominate
fig, axes = plt.subplots(1, 2, figsize=(12, 5), sharex=True, sharey=True)
axes = axes.ravel()

# Compute global y-limits
lat_min = df[df["scenario"].isin(SCENARIOS)]["latency_ms"].min()
lat_max = df[df["scenario"].isin(SCENARIOS)]["latency_ms"].max()

title = ["Decode Only", "B-1 Decode + 1 Prefill"]
for ax, scenario, t in zip(axes, SCENARIOS, title):
    sub = df[df["scenario"] == scenario]

    # rows = seq_len, cols = batch size
    piv_meas = sub.pivot_table(index="seq_len", columns="batch", values="latency_ms")

    for B in sorted(piv_meas.columns):
        ax.plot(piv_meas.index, piv_meas[B],
                marker="o", ls="-", label=f"B={B}")

    ax.set_xscale("log", base=2)
    ax.set_ylim(lat_min * 0.95, lat_max * 1.05)  # uniform y-limits with margin
    ax.set_title(t)
    ax.set_xlabel("Sequence length")
    ax.grid(True, ls="--", lw=0.5)
    ax.legend(fontsize=8)

# Shared y-axis label
fig.text(0.04, 0.5, "Latency per batch (ms)", va="center", rotation="vertical")

plt.tight_layout(rect=[0.04, 0.04, 1, 1])
plt.savefig("fa3_latency_debug.png", dpi=150)
```

**Comment by leigao97** [comment]
Hi @tridao, just wanted to follow up with you. Were you able to reproduce the issue? Thank you.
